File: Backend/app/pw_library/python_methods/lagrange.py
import sympy as sp


def lagrange_polynomial(x_values, y_values):
    """
    Compute the Lagrange polynomial for a given set of data points.

    Parameters:
    x_values (list or array-like): List of x-coordinates of the data points.
    y_values (list or array-like): List of y-coordinates of the data points.

    Returns:
    sympy.Expr: The Lagrange polynomial as a symbolic expression.
    """
    if len(x_values) != len(y_values):
        raise ValueError("x_values and y_values must have the same length.")

    # Define the symbolic variable for the polynomial
    x = sp.symbols('x')

    # Initialize the polynomial
    polynomial = 0

    # Construct the Lagrange polynomial
    n = len(x_values)
    for i in range(n):
        # Start with the y value for this term
        term = y_values[i]

        # Multiply by each Lagrange basis polynomial
        for j in range(n):
            if j != i:
                term *= (x - x_values[j]) / (x_values[i] - x_values[j])

        # Add this term to the polynomial
        polynomial += term

    # Simplify the polynomial expression
    return sp.simplify(polynomial)

File: Backend/app/pw_library/rust_methods/lagrange.rs
use ndarray::{Array1, Array2, ArrayView1};
use num_complex::Complex64;
use sympy::{Symbol, sympify, Expr, simplify};

fn lagrange_polynomial(x_values: &ArrayView1<f64>, y_values: &ArrayView1<f64>) -> Expr {
    if x_values.len() != y_values.len() {
        panic!("x_values and y_values must have the same length.");
    }

    let n = x_values.len();
    let x = Symbol::from("x");
    let mut polynomial = Expr::from(0);

    for i in 0..n {
        let mut term = y_values[i];
        for j in 0..n {
            if j != i {
                term *= (x.clone() - x_values[j]) / (x_values[i] - x_values[j]);
            }
        }
        polynomial += term;
    }

    simplify(polynomial)
}